extensions/infinispan-cache/runtime/src/main/java/io/quarkus/cache/infinispan/runtime/InfinispanCacheImpl.java

```diff
@@ -0,0 +1,215 @@
+                return computedValue;
+            }
+            V newValue = valueLoader.apply(key);
+            remoteCache
+                    .putIfAbsentAsync(key, encodeNull(newValue), lifespan, TimeUnit.MILLISECONDS, maxIdle,
+                            TimeUnit.MILLISECONDS)
+                    .whenComplete((existing, ex2) -> {
+                        if (ex2 != null) {
+                            resultAsync.completeExceptionally((Throwable) ex2);
+                        } else if (existing == null) {
+                            resultAsync.complete(newValue);
+                        } else {
+                            resultAsync.complete(decodeNull(existing));
+                        }
+                        computationResults.remove(key);
+                    });
+            return resultAsync;
+        }));
+    }
+
+    @Override
+    public <K, V> Uni<V> getAsync(K key, Function<K, Uni<V>> valueLoader) {
+        return Uni.createFrom().completionStage(handleAndCompose(remoteCache.getAsync(key), (v1, ex1) -> {
+            if (ex1 != null) {
+                return CompletableFuture.failedFuture(ex1);
+            }
+
+            if (v1 != null) {
+                return CompletableFuture.completedFuture(decodeNull(v1));
+            }
+
+            CompletableFuture<V> resultAsync = new CompletableFuture<>();
+            CompletableFuture<V> computedValue = computationResults.putIfAbsent(key, resultAsync);
+            if (computedValue != null) {
+                return computedValue;
+            }
+            valueLoader.apply(key).convert().toCompletionStage()
+                    .whenComplete((newValue, ex2) -> {
+                        if (ex2 != null) {
+                            resultAsync.completeExceptionally(ex2);
+                            computationResults.remove(key);
+                        } else {
+                            remoteCache.putIfAbsentAsync(key, encodeNull(newValue), lifespan, TimeUnit.MILLISECONDS, maxIdle,
+                                    TimeUnit.MILLISECONDS).whenComplete((existing, ex3) -> {
+                                        if (ex3 != null) {
+                                            resultAsync.completeExceptionally((Throwable) ex3);
+                                        } else if (existing == null) {
+                                            resultAsync.complete(newValue);
+                                        } else {
+                                            resultAsync.complete(decodeNull(existing));
+                                        }
+                                        computationResults.remove(key);
+                                    });
+                        }
+                    });
+            return resultAsync;
+        }));
+    }
+
+    @Override
+    public Uni<Void> invalidate(Object key) {
+        return Uni.createFrom().completionStage(remoteCache.removeAsync(key));
```

Comment:
That is not correct.
This invokes the "removeAsync" method immediately, and not at subscription time. 

You need `() -> removeCache.removeAsync(key))` (for all methods)

In addition, you may have to deal with the context, as your event is going to be emitted on a  infinispan thread, which is not the caller context. At least add tests (the interceptor is likely going to do that already, but better verify).

Reply:
There is a new integration test, what else should I add to test about the context ?